### PR TITLE
#12293: Migrate moreh_arange operation from tt_eager to ttnn

### PR DESCRIPTION
**Pull request by mrshaw01**
### Ticket
Link to Github Issue: [#12293](https://github.com/tenstorrent/tt-metal/issues/12293)

### Problem description
`moreh_arange` was deprecated alongside `tt_dnn`. This PR ports it to `ttnn` using its operation format.

### What's changed
- Removed `moreh_arange`'s `tt_dnn` implementations
- Move device code to `ttnn`
- Create new wrapper code for `ttnn` with new modules

### Checklist
- [x] Post commit CI passes: 
+ [All post-commit tests #15078](https://github.com/tenstorrent/tt-metal/actions/runs/10733410773)
+ [All post-commit tests #15094](https://github.com/tenstorrent/tt-metal/actions/runs/10735740160)
+ [All post-commit tests #15098](https://github.com/tenstorrent/tt-metal/actions/runs/10736261451)
+ [All post-commit tests #15113](https://github.com/tenstorrent/tt-metal/actions/runs/10739414437)
+ [All post-commit tests #15189](https://github.com/tenstorrent/tt-metal/actions/runs/10750368234)
+ [All post-commit tests #15308](https://github.com/tenstorrent/tt-metal/actions/runs/10786219800)
- [x] Blackhole Post commit (if applicable): [Blackhole post-commit tests #521](https://github.com/tenstorrent/tt-metal/actions/runs/10735773734)
- [x] Model regression CI testing passes (if applicable): Not available.
- [x] New/Existing tests provide coverage for changes: 18/18 test cases passed.

### CODE REVIEW

**Comment by mrshaw01** [comment]
Running Actions:
- [All post-commit tests #15094](https://github.com/tenstorrent/tt-metal/actions/runs/10735740160)
- [Blackhole post-commit tests #521](https://github.com/tenstorrent/tt-metal/actions/runs/10735773734)

File: ttnn/cpp/ttnn/operations/moreh/moreh_arange/moreh_arange.hpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        const std::optional<Tensor>& output_tensor,
+        bool untilize_out,
+        const std::optional<DataType>& output_dtype,
+        const std::optional<MemoryConfig>& output_memory_config);
+};
+}  // namespace ttnn::operations::moreh::moreh_arange
+
+namespace ttnn {
+constexpr auto moreh_arange =
+    ttnn::register_operation<"ttnn::moreh_arange", ttnn::operations::moreh::moreh_arange::MorehArange>();
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_arange/moreh_arange.hpp:26]:
```suggestion
    ttnn::register_operation_with_auto_launch_op<"ttnn::moreh_arange", ttnn::operations::moreh::moreh_arange::MorehArange>();
```
Composite ops should use `register_operation_with_auto_launch_op`

File: ttnn/cpp/ttnn/operations/moreh/moreh_arange/device/moreh_arange_device_operation.hpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+namespace ttnn::operations::moreh::moreh_arange {
+
+struct MorehArangeOperation {
+    struct operation_attributes_t {
+        float start;
+        float end;
+        float step;
+        bool untilize_out;
+        const std::optional<DataType> output_dtype;
+        const std::optional<MemoryConfig> output_memory_config;
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_arange/device/moreh_arange_device_operation.hpp:22]:
```suggestion
        const MemoryConfig output_memory_config;
```
 Inside operation_attributes_t, it’s MemoryConfig.

Comment (mrshaw01) [reply ttnn/cpp/ttnn/operations/moreh/moreh_arange/device/moreh_arange_device_operation.hpp]: I changed it as TT's method.

**Review by hschoi4448 — APPROVED** [review]

**Review by razorback3 — APPROVED** [review]

**Comment by mrshaw01** [comment]
@ayerofieiev-tt I think the Moreh-side review is done. Could you please review my PR?

File: ttnn/cpp/ttnn/operations/moreh/moreh_arange/device/moreh_arange_device_operation.cpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+#include "ttnn/tensor/tensor.hpp"
+
+namespace ttnn::operations::moreh::moreh_arange {
+void MorehArangeOperation::validate_inputs(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    auto step = operation_attributes.step;
+    auto start = operation_attributes.start;
+    auto end = operation_attributes.end;
+    TT_ASSERT(step != 0, "step must be nonzero.");
+    TT_ASSERT(
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/moreh/moreh_arange/device/moreh_arange_device_operation.cpp:19]: TT_ASSERT only happens in debug builds. 
Please use TT_FATAL for user input validation for consistency.

File: ttnn/cpp/ttnn/operations/moreh/moreh_arange/moreh_arange.hpp [hunk 4]
```diff
@@ -0,0 +1,27 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "ttnn/decorators.hpp"
+#include "ttnn/operations/moreh/moreh_arange/device/moreh_arange_device_operation.hpp"
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/moreh/moreh_arange/moreh_arange.hpp:8]: This include is not needed here, right?

File: ttnn/cpp/ttnn/operations/moreh/moreh_arange/moreh_arange_pybind.cpp [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        ttnn::moreh_arange,
+        "Moreh Arange Operation",
+        ttnn::pybind_arguments_t{
+            py::arg("start"),
+            py::arg("end"),
+            py::arg("step"),
+            py::arg("any"),
+            py::arg("output_tensor") = std::nullopt,
+            py::arg("untilize_out") = false,
+            py::arg("output_dtype") = std::nullopt,
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/moreh/moreh_arange/moreh_arange_pybind.cpp:23]: In ttnn we use kw_only arguments dtype and memory_config. 
please adjust for consistency. 
Also make sure that arguments which are kw args in torch are also kw args here. 

Thank you!

**Review by ayerofieiev-tt — APPROVED** [review]
Overall looks good. 
Please address the feedback, run post commit workflow and proceed with merge.

**Comment by mrshaw01** [comment]
The current situation is that all post-commit tests have failed on the main branch. Consequently, the post-commit tests for this branch (which is up to date with the main branch) also fail on the exact same test cases. I also found another branch experiencing the same situation.
- [All post-commit tests #15308](https://github.com/tenstorrent/tt-metal/actions/runs/10786219800/job/29913387240)
- [#6721: Log nan fix (#12167) #15305](https://github.com/tenstorrent/tt-metal/actions/runs/10785847518/job/29914324764)
- [All post-commit tests #15309](https://github.com/tenstorrent/tt-metal/actions/runs/10786494728)

**Pull request merged**